`LineProfilerMagics` refactoring, coverage boost, other fixes

TTsangSC opened the pull request:
Code changes
---
- Refactored `line_profiler/ipython_extensions.py::LineProfilerMagics.lprun()` and `.lprun_all()` to unify code paths
- Fixed error message written by `.lprun` ("Could not find module [...]" -> "Could not find **function** [...]") which an `-f` target can't be profiled
- Fixed edge case where `.lprun_all()` gets a blank `cell`
- Fixed misuse of cell magic (`%lprun_all` -> `%%lprun_all`) in `.lprun_all.__doc__`
- No longer enclosing the profiled code in a dummy function. This has the following benefits:
  - Fixing bug where local names defined in the `%%lprun_all`-ed code block are not available in the subsequent namespace
  - Making profiling output cleaner; instead of
    ```
    [...]  Line Contents
    ====================
    [...]  def _lprun_cell():
    [...]      content
    ```
    we now just have
    ```
    [...]  Line Contents
    ====================
    [...]  content
    ```
- Removed the now-unused `SkipWrapper`
- Updated docstring formats to be more Sphinx-friendly

Test changes
---
- Refactored `tests/test_ipython.py::TestIPython` into the following test classes:
  - `TestLPRun`:  
    - Tests for `%lprun`, its specific options (`-m`, `-s`), and options common to `%lprun` and `%%lprun_all` (`-D`, `-T`, `-u`, `-z`)
    - Supersedes `TestIPython.test_init()`
  - `TestLPRunAll`:  
    - Tests for `%%lprun_all`, migrated from `TestIPython.test_lprun_all_*()`
    - Added check that cell namespace changes are correctly handled
- File coverage for `line_profiler.ipython_extension` is now at 98% compared to the previous 70-something

TTsangSC committed:
Stub

TTsangSC committed:
Consolidate common functionalities

- `line_profiler/ipython_extension.py[i]`
  - General:
    Added type comments for sanity checks
  - `_ParseParamResult`:
    New private class summarizing the result of argument-parsing
  - `_RunAndProfileResult`:
    New private class summarizing the result of calling a function with
    profiling
  - `_PatchProfilerIntoBuiltins`:
    New private helper context-manager class for monkey-patching a
    profiler into `builtins`
  - `LineProfilerMagics`:
    - `_parse_parameters()`:
      - Refactored from `parse_parameters()`, taking over more of the
        postprocessing boilerplate
      - Now returning a `_ParseParamResult`
    - `_run_and_profile()`:
      New private method to call a callable, profile and time the call,
      and handle exception-related boilerplate;
      returns a `_RunAndProfileResult`
    - `_lprun_get_top_level_profiled_code()`,
      `_lprun_all_get_rewritten_profiled_code()`:
      New private helper methods for handling the compilation of code in
      the `%%lprun_all -p` and `%%lprun_all` cases
    - `_handle_end()`:
      New private method to handle end-of-magic boilerplate (IO, return
      values)
    - `lprun()`:
      - Simplified implementation to use the above helper methods
      - Fixed erroneous error message when a function to be profiled is
        not found ("Could not find module [...]"
        -> "Could not find function [...]")
    - `lprun_all()`:
      - Simplified implementation to use the above helper methods
      - Fixed misuse of cell magic in docstring ("%lprun_all [...]"
        -> "%%lprun_all [...]")
      - Fixed edge case where the cell is blank
      - Now ensuring that the profiling output is formatted before
        deleting tempfiles

TTsangSC committed:
More Sphinx-friendly docs

TTsangSC committed:
Prettier output and scoping fix

- `line_profiler/ipython_extension.py`:
  - `_RunAndProfileResult`:
    - `tempfile`:
      New attribute
    - `output`:
      Updated implementation to allow for module- (i.e. top-) level
      profiling data (from the code object `.tempfile` compiles into) to
      be formatted correctly
    - `_make_show_func_wrapper()`:
      New private helper method for creating a monkey-patched clone of
      `line_profiler.line_profiler.show_func()` which implements the
      above
  - `_PatchProfilerIntoBuiltins`:
    Simplified implementation to use a `_PatchDict` internally
  - `_PatchDict`:
    New private comtext-manager helper class for monkey-patching
    arbitrary namespaces
  - `LineProfilerMagics`:
    - `_run_and_profile()`:
      Added argument `tempfile`, which is passed to
      `_RunAndProfileResult`
    - `_lprun_all_get_rewritten_profiled_code()`:
      No longer uses `SkipWrapper`
    - `lprun_all()`:
      - Removed nesting of the `cell` content into a mock function;
        the content is now directly written to the tempfile, compiled,
        and executed, allowing for prettier output
      - Fixed scoping of names: local names in `cell` are now correctly
        reflected in `IPython.core.getipython.get_ipython().user_ns`
- `tests/test_ipython.py::TestIPython`
  - `loops`, `lprun_all_cell_body`:
    Extracted common assets into class attributes
  - `test_lprun_all_*()`:
    - Removed line-number offsets induced by the nesting of the cell
      content into a mock function
    - Added checks that the cell content is executed in the correct
      scope

TTsangSC committed:
Boost coverage

- `line_profiler/ipython_extension.py`
  - General:
    Added `# pragma: no cover` to type-checking only stuff
  - Removed the following unused code:
    - `SkipWrapper`
    - `_ParseParamResult.__getitem__()`
    - `_RunAndProfileResult.return_value`

- `tests/test_ipython.py`
  - `TestIPython`:
    Refactored into the classes `TestLPRun` and `TestLPRunAll`, which
    are no longer `unittest.TestCase` subclasses
  - `TestIPython.test_init()`:
    Refactored and parametrized into the following tests:
    - `test_lprun_profiling_targets()`:
      Test `-m`
    - `test_lprun_file_io()`:
      Test `-D` and `-T`
    - `test_lprun_timer_unit()`:
      Test `-u`
    - `test_lprun_skip_zero()`:
      Test `-s` and `-z`
    - `test_lprun_exception_handling()`:
      Test exception handling in the magics
  - `TestLPRunAll.test_lprun_all_*()`:
    Updated implementations

Theelx commented on line_profiler/ipython_extension.py:
```diff
@@ -227,279 +434,115 @@ def lprun(self, parameter_s=""):
-        -z: strip out all entries from the print-out that have zeros. Note: this is -s in
-        lprun, however we use -z here for consistency with the CLI.
+        ``-z``: strip out all entries from the print-out that have
+        zeros.
```
Can we include the part where -z is used here for consistency with the CLI?

Theelx reviewed (approved):
This looks great, thanks! While I do indeed want to get this approved sooner rather than later, this is such a massive improvement over the existing code that I think it's worth merging in wholesale and merging into the PR before my existing PR becomes mainline, even if it takes more review time. I am a bit worried about the maintainability of this code by a human because of all the indirection to special classes, however if you think that's all necessary then I'll put in additional effort to understand it better!

Theelx committed:
Fix grammar in documentation

One space after a period is generally considered correct. Also, add a clarification that the -s option is meant to eventually be deprecated, and that -z is used for consistency with the CLI.
